Implement beat

crash-g opened the pull request:
This is a draft for the implementation of the celery beat ([issue 55](https://github.com/rusty-celery/rusty-celery/issues/55)). A usage example can be found in `examples/beat_app.rs`. Changes to the rest of the project were kept as small as possible.

Possible next steps:
- clean up the `beat` module;
- accept task/schedule pairs in input in the `beat!` macro;
- create a `BeatService` builder, reusing logic from `CeleryBuilder` where possible.

epwalsh commented:
Thanks @crash-g, looks like a great start!

crash-g commented on src/protocol/mod.rs:
```diff
@@ -193,18 +193,29 @@ where
 
+/// A trait for attempting to deserialize a `Message` from `self`. This is required to be implemented
+/// on a broker's `Delivery` type.
+pub trait TryDeserializeMessage {
```
This trait was previously called `TryCreateMessage`. I have changed its name trying to highlight this difference:
- the `TryCreateMessage` is a message factory which creates new messages to be sent
- the `TryDeserializeMessage` is used to deserialize a received message in order to execute it

epwalsh commented:
@crash-g looks like there's a few merge conflicts, could you fix those when you get a chance?

crash-g commented:
> @crash-g looks like there's a few merge conflicts, could you fix those when you get a chance?

Done. Please double check that the implementation of `build_and_connect` in the broker module is correct following the merge.

epwalsh commented:
Sorry I've been slow to get back to you @crash-g. Could you fix the CI errors?

crash-g commented:
> Sorry I've been slow to get back to you @crash-g. Could you fix the CI errors?

I have fixed format and linting, and I've also added documentation. The only check which is failing now is the changelog, but I want to update it as last thing before merging the branch.

A couple of things to consider before merging:
1. Should the beat example be made more readable?
2. What tests to add?

epwalsh commented:
Thanks @crash-g! Is there a good way to add an integration test? Ideally we'd have one for each broker like in https://github.com/rusty-celery/rusty-celery/tree/master/tests/brokers. I don't if that's feasible though.

epwalsh commented on examples/beat_app.rs:
```diff
@@ -0,0 +1,73 @@
+    );
+
+    // Build a `Beat` with a custom scheduler backend.
+    let custom_backend_beat = celery::beat!(
```
I think this example (of using custom backend) could just be put in the documentation and left out of the example. Just so we keep the example as simple as possible

crash-g replied:
Done

epwalsh commented:
As far as simplifying the example, what if we just moved the beat example into the existing `examples/celery_app.rs`? Then to run the beat producer you could just do `cargo run --example celery_app beat` or something? That way there's less duplication in the examples, so it might make it a little easier to maintain. Then again, if that makes it too complicated then it's not worth it.

crash-g commented:
> As far as simplifying the example, what if we just moved the beat example into the existing `examples/celery_app.rs`? Then to run the beat producer you could just do `cargo run --example celery_app beat` or something? That way there's less duplication in the examples, so it might make it a little easier to maintain. Then again, if that makes it too complicated then it's not worth it.

I have modified the `beat_app` example to work with the existing consumers, so `beat_consumer` could be removed. I think this solution is clean enough (I have updated the README to describe how to run the examples).

crash-g commented:
> Thanks @crash-g! Is there a good way to add an integration test? Ideally we'd have one for each broker like in https://github.com/rusty-celery/rusty-celery/tree/master/tests/brokers. I don't if that's feasible though.

I have experimented with a [unit test](https://github.com/crash-g/rusty-celery/blob/master/src/beat/tests.rs) which uses a dummy broker. I think that the test is effective but:
1. it requires a little boilerplate
2. it is time-sensitive, so in (hopefully unlikely) scenarios may fail

If we don't like 1 or 2 we may think of another test strategy.

By the way, current CI failure is unrelated (I will fix it ASAP).

epwalsh commented:
This should fix a small bug a noticed with the connection logic: https://github.com/crash-g/rusty-celery/pull/1

epwalsh commented on src/beat/scheduler.rs:
```diff
@@ -0,0 +1,136 @@
+
+        match scheduled_task.message_factory.try_create_message() {
+            Ok(message) => {
+                debug!("Sending task {} to {} queue", scheduled_task.name, queue);
```
I think it would nice to log this at INFO, as it is with Python.

```suggestion
                info!("Sending task {} to {} queue", scheduled_task.name, queue);
```

crash-g replied:
Fixed

epwalsh commented on src/broker/mod.rs:
```diff
@@ -89,3 +94,69 @@ pub trait BrokerBuilder {
+// TODO: this function consumes the broker_builder, which results in a not so ergonomic API.
+// Can it be improved?
+/// A utility function to configure the task routes on a broker builder.
+pub(crate) fn configure_task_routes<Bb: BrokerBuilder>(
```
We might want to just have a separate `utils` module or something. Not a high priority though.

crash-g replied:
I'll leave it there for now but feel free to move it somewhere else later on

crash-g commented on src/beat/mod.rs:
```diff
@@ -0,0 +1,276 @@
+/// created with the [`beat`](macro.beat.html) macro.
+///
+/// The *beat* is in charge of executing scheduled tasks when
+/// they are due and add or remove tasks as required. It drives execution by
```
Fixed

crash-g commented:
> My only significant comments are about error handling. I think we should propagate all beat errors from the `Broker`, `Scheduler`, and `SchedulerBackend` up to the Beat app (in particular, the `Beat::start` method), just like we do with `Celery::consume` for example.

I have provided two methods:
- `try_start` has the signature you suggested (it stops in case of errors)
- `start` has the same signature as before (it only logs errors when they happen)

crash-g commented:
> So, I still think `start()` should return errors instead of looping forever

I agree with you, with the latest commit I have left only a `start()` method which returns errors as suggested.

> Also, could you edit this line in the README to show that we now support Beat? https://github.com/rusty-celery/rusty-celery/pull/127/files#diff-04c6e90faac2675aa89e2176d2eec7d8R152

Done!
